File: source/addonManager51/Views/collectionLibraryClass.vb

Imports Contensive.Addons.PortalFramework
Imports Contensive.BaseClasses

Namespace Contensive.Addons.AddonManager51
    '
    ' Sample Vb addon
    '
    Public Class collectionLibraryClass
        Inherits AddonBaseClass
        '
        ' injected objects -- do not dispose
        '
        Private cp As CPBaseClass
        ' 
        ' class scope
        '
        Private Const RequestNameButton As String = "button"
        Private Const CollectionListRootNode As String = "collectionlist"
        '
        Private Const ButtonCancel As String = " Cancel "
        Private Const ButtonOK As String = " OK "
        '
        Private Structure NavigatorType
            Public Name As String
            Public NameSpacex As String
        End Structure

        Private Structure Collection2Type
            Public AddonCnt As Integer
            Public AddonGuid() As String
            Public AddonName() As String
            Public MenuCnt As Integer
            Public Menus() As String
            Public NavigatorCnt As Integer
            Public Navigators() As NavigatorType
        End Structure
        '
        Private CollectionCnt As Integer
        Private Collections() As Collection2Type
        Private Const guidAddonManagerLibraryListCell = "{9767F464-3728-4B7D-904B-3442D7FD03BE}"
        Private Const guidAddonManagerActiveX = "{1DC06F61-1837-419B-AF36-D5CC41E1C9FD}"
        '
        '=====================================================================================
        ' addon api
        '=====================================================================================
        '
        Public Overrides Function Execute(ByVal CP As CPBaseClass) As Object
            Dim returnHtml As String = ""
            Try
                Me.cp = CP
                returnHtml = GetCollectionLibrary()
            Catch ex As Exception
                CP.Site.ErrorReport(ex)
            End Try
            Return returnHtml
        End Function
        '
        '==========================================================================================================================================
        '   Addon Manager
        '       This is a form that lets you upload an addon
        '       Eventually, this should be substituted with a "Addon Manager Addon" - so the interface can be improved with Contensive recompile
        '==========================================================================================================================================
        '
        Private Function GetCollectionLibrary() As String
            Dim returnResult As String = ""
            Try
                Dim form As New FormSimpleClass
                Dim showAddon As Boolean
                Dim DbUpToDate As Boolean
                Dim GuidFieldName As String
                Dim ErrorMessage As String = ""
                Dim UpgradeOK As Boolean
                Dim LibCollections As Xml.XmlDocument
                Dim installFolder As String
                Dim LibGuids() As String
                Dim CollectionGUID As String = ""
                Dim CollectionVersion As String
                Dim CollectionContensiveVersion As String = ""
                Dim cnt As Integer
                Dim Ptr As Integer
                Dim RowPtr As Integer
                Dim PageNumber As Integer
                Dim ColumnCnt As Integer
                Dim ColCaption() As String
                Dim ColAlign() As String
                Dim ColWidth() As String
                Dim ColSortable() As Boolean
                Dim PreTableCopy As String
                Dim BodyHTML As String = ""
                Dim UserError As String
                Dim Button As String
                Dim ButtonList As String
                Dim CollectionFilename As String = ""
                Dim Doc As New Xml.XmlDocument
                Dim CollectionNode As Xml.XmlNode
                Dim status As String = ""
                Dim collectionsToBeInstalledFromFolder As Boolean
                Dim InstallLibCollectionList As String = ""
                Dim InstallPath As String
                Dim SiteKey As String
                '
                SiteKey = cp.Site.GetText("sitekey", "")
                If SiteKey = "" Then
                    SiteKey = cp.Utils.CreateGuid()
                    Call cp.Site.SetProperty("sitekey", SiteKey)
                End If
                '
                DbUpToDate = (cp.Site.GetText("buildVersion") >= cp.Version)
                Button = cp.Doc.GetText(RequestNameButton)
                collectionsToBeInstalledFromFolder = False
                GuidFieldName = "ccguid"
                If Button = ButtonCancel Then
                    '
                    ' ----- redirect back to the root
                    '
                    Call cp.Response.Redirect(cp.Site.GetText("adminUrl"))
                Else
                    If Not cp.User.IsAdmin() Then
                        '
                        ' ----- Put up error message
                        '
                        ButtonList = ButtonCancel
                        BodyHTML = cp.Html.p("You must be an administrator to use this tool.")
                    Else
                        '
                        PreTableCopy = "Use this form to upload an add-on collection. If the GUID of the add-on matches one already installed on this server, it will be updated. If the GUID is new, it will be added."
                        installFolder = "CollectionUpload" & cp.Utils.CreateGuid().Replace("{", "").Replace("-", "").Replace("}", "")
                        InstallPath = installFolder & "\"
                        If (Button = ButtonOK) Then
                            '
                            '---------------------------------------------------------------------------------------------
                            ' Download and install Collections from the Collection Library
                            '---------------------------------------------------------------------------------------------
                            '
                            InstallLibCollectionList = ""
                            If cp.Doc.GetText("LibraryRow") <> "" Then
                                Ptr = cp.Doc.GetInteger("LibraryRow")
                                CollectionGUID = cp.Doc.GetText("LibraryRowguid" & Ptr)
                                InstallLibCollectionList = InstallLibCollectionList & "," & CollectionGUID
                            End If
                            '
                            If InstallLibCollectionList <> "" Then
                                InstallLibCollectionList = Mid(InstallLibCollectionList, 2)
                                LibGuids = Split(InstallLibCollectionList, ",")
                                cnt = UBound(LibGuids) + 1
                                For Ptr = 0 To cnt - 1
                                    '
                                    ' -- use v5 method
                                    UpgradeOK = v5InstallController.installCollectionFromLibrary(cp, LibGuids(Ptr), ErrorMessage)
                                Next
                            End If
                        End If
                        '
                        ' --------------------------------------------------------------------------------
                        ' Get Form
                        ' --------------------------------------------------------------------------------
                        ' Get the Collection Library tab
                        ' --------------------------------------------------------------------------------
                        '
                        ColumnCnt = 4
                        PageNumber = 1
                        ReDim ColCaption(3)
                        ReDim ColAlign(3)
                        ReDim ColWidth(3)
                        ReDim ColSortable(3)
                        '
                        ColCaption(0) = "Install"
                        ColAlign(0) = "center"
                        ColWidth(0) = "50"
                        ColSortable(0) = False
                        '
                        ColCaption(1) = "Name"
                        ColAlign(1) = "left"
                        ColWidth(1) = "200"
                        ColSortable(1) = False
                        '
                        ColCaption(2) = "Last&nbsp;Updated"
                        ColAlign(2) = "right"
                        ColWidth(2) = "200"
                        ColSortable(2) = False
                        '
                        ColCaption(3) = "Description"
                        ColAlign(3) = "left"
                        ColWidth(3) = "99%"
                        ColSortable(3) = False
                        '
                        LibCollections = New Xml.XmlDocument
                        Call LibCollections.Load("http://support.contensive.com/GetCollectionList?iv=" & cp.Version & "&key=" & cp.Utils.EncodeRequestVariable(SiteKey) & "&name=" & cp.Utils.EncodeRequestVariable(cp.Site.Name) & "&primaryDomain=" & cp.Utils.EncodeRequestVariable(cp.Site.DomainPrimary))
                        If True Then
                            If (LibCollections.DocumentElement.Name.ToLower() <> CollectionListRootNode.ToLower()) Then
                                UserError = "There was an error reading the Collection Library file. The '" & CollectionListRootNode & "' element was not found."
                                status = status & "<BR>" & UserError
                                cp.UserError.Add(UserError)
                            Else
                                '
                                ' Go through file to validate the XML, and build status message -- since service process can not communicate to user
                                '
                                RowPtr = 0
                                'Content = ""
                                Dim cellTemplate As String = My.Resources.AddonManagerLibraryListCell
                                For Each CDef_Node As Xml.XmlNode In LibCollections.DocumentElement.ChildNodes
                                    Dim Cell As String = cellTemplate
                                    Dim CollectionImageLink As String = ""
                                    Dim CollectionCheckbox As String = ""
                                    Dim CollectionName As String = ""
                                    Dim CollectionModifiedDate As Date = Date.MinValue
                                    Dim CollectionModifiedDateCaption As String = ""
                                    Dim CollectionDescription As String = ""
                                    Dim CollectionHelpLink As String = ""
                                    Dim CollectionDemoLink As String = ""
                                    Select Case LCase(CDef_Node.Name)
                                        Case "collection"
                                            '
                                            ' Read the collection
                                            '
                                            For Each CollectionNode In CDef_Node.ChildNodes
                                                Select Case CollectionNode.Name.ToLower()
                                                    Case "name"
                                                        '
                                                        ' Name
                                                        '
                                                        CollectionName = CollectionNode.InnerText
                                                    Case "helplink"
                                                        '
                                                        ' helpLink
                                                        '
                                                        CollectionHelpLink = CollectionNode.InnerText
                                                    Case "demolink"
                                                        '
                                                        ' demoLink
                                                        '
                                                        CollectionDemoLink = CollectionNode.InnerText
                                                    Case "guid"
                                                        '
                                                        ' Guid
                                                        '
                                                        CollectionGUID = CollectionNode.InnerText
                                                    Case "version"
                                                        '
                                                        ' Version
                                                        '
                                                        CollectionVersion = CollectionNode.InnerText
                                                    Case "description"
                                                        '
                                                        ' Version
                                                        '
                                                        CollectionDescription = CollectionNode.InnerText
                                                    Case "imagelink"
                                                        '
                                                        ' Version
                                                        '
                                                        CollectionImageLink = CollectionNode.InnerText
                                                    Case "contensiveversion"
                                                        '
                                                        ' Version
                                                        '
                                                        CollectionContensiveVersion = CollectionNode.InnerText
                                                    Case "lastchangedate"
                                                        '
                                                        ' last change - legacy field used by 4.1 to auto install, no longer updated
                                                        '
                                                    Case "lastmodifieddate"
                                                        '
                                                        ' last modified
                                                        '
                                                        CollectionModifiedDate = Date.MinValue
                                                        If IsDate(CollectionNode.InnerText) Then
                                                            CollectionModifiedDate = CDate(CollectionNode.InnerText)
                                                        End If
                                                        If (CollectionModifiedDate <= Date.MinValue) Then
                                                            CollectionModifiedDateCaption = "unknown"
                                                        Else
                                                            CollectionModifiedDateCaption = CollectionModifiedDate.Date.ToShortDateString
                                                        End If
                                                End Select
                                            Next
                                            If CollectionImageLink = "" Then
                                                CollectionImageLink = "/addonManager/libraryNoImage.jpg"
                                            End If
                                            If CollectionModifiedDateCaption = "" Then
                                                CollectionModifiedDateCaption = "unknown"
                                            End If
                                            If CollectionDescription = "" Then
                                                CollectionDescription = "No description is available for this add-on collection."
                                            End If
                                            showAddon = False
                                            If CollectionName = "" Then
                                            Else
                                                If CollectionGUID = "" Then
                                                Else
                                                    Dim cs As CPCSBaseClass = cp.CSNew
                                                    Dim modifiedDate As Date = Date.MinValue
                                                    Dim IsOnSite As Boolean = False
                                                    If cs.Open("Add-on Collections", GuidFieldName & "=" & cp.Db.EncodeSQLText(CollectionGUID), "", True, "ID,ModifiedDate") Then
                                                        modifiedDate = cs.GetDate("ModifiedDate")
                                                        IsOnSite = True
                                                    End If
                                                    Call cs.Close()
                                                    CollectionCheckbox = "<input TYPE=""CheckBox"" NAME=""LibraryRow"" VALUE=""" & RowPtr & """ onClick=""clearLibraryRows('" & RowPtr & "');"">" & cp.Html.Hidden("LibraryRowGuid" & RowPtr, CollectionGUID) & cp.Html.Hidden("LibraryRowName" & RowPtr, CollectionName)
                                                    If (Not IsOnSite) Then
                                                        '
                                                        ' -- not installed
                                                        showAddon = True
                                                        CollectionCheckbox &= "&nbsp;Install"
                                                    ElseIf (modifiedDate >= CollectionModifiedDate) Then
                                                        '
                                                        ' -- up to date, reinstall
                                                        showAddon = True
                                                        CollectionCheckbox &= "&nbsp;Reinstall"
                                                    Else
                                                        '
                                                        ' -- old version, upgrade
                                                        showAddon = True
                                                        CollectionCheckbox &= "&nbsp;Upgrade"
                                                    End If
                                                End If
                                            End If
                                            If CollectionDemoLink <> "" Then
                                                CollectionDescription = CollectionDescription & "<div class=""amDemoLink""><a target=""_blank"" href=""" & CollectionDemoLink & """>Demo</a></div>"
                                            End If
                                            If CollectionHelpLink <> "" Then
                                                CollectionDescription = CollectionDescription & "<div class=""amHelpLink""><a target=""_blank"" href=""" & CollectionHelpLink & """>Reference</a></div>"
                                            End If
                                            If showAddon Then
                                                Cell = Replace(Cell, "##imageLink##", CollectionImageLink)
                                                Cell = Replace(Cell, "##checkbox##", CollectionCheckbox)
                                                Cell = Replace(Cell, "##name##", CollectionName)
                                                Cell = Replace(Cell, "##date##", CollectionModifiedDateCaption)
                                                Cell = Replace(Cell, "##description##", CollectionDescription)
                                                BodyHTML = BodyHTML & Cell
                                            End If
                                            RowPtr = RowPtr + 1
                                    End Select
                                Next
                            End If
                            BodyHTML = "" _
                            & cr & "<script language=""JavaScript"">" _
                            & "function clearLibraryRows(r) {" _
                            & "var c,p;" _
                            & "c=document.getElementsByName('LibraryRow');" _
                                & "for (p=0;p<c.length;p++){" _
                                    & "if(c[p].value!=r)c[p].checked=false;" _
                                & "}" _
                            & "" _
                            & "}" _
                            & "</script>" _
                            & "<input type=hidden name=LibraryCnt value=""" & RowPtr & """>" _
                            & cr & "<div style=""width:100%"">" _
                            & kmaIndent(BodyHTML) _
                            & cr & "</div>" _
                            & ""
                            form.body = BodyHTML
                            form.description = "Select Add-ons to install from the Contensive Add-on Library. Please select only one at a time. Click OK to install the selected Add-on. The site may need to be stopped during the installation, but will be available again in approximately one minute"

                            'BodyHTML = AdminUI.GetEditPanel(main, True, "Add-on Collection Library", "Select Add-ons to install from the Contensive Add-on Library. Please select only one at a time. Click OK to install the selected Add-on. The site may need to be stopped during the installation, but will be available again in approximately one minute.", BodyHTML)
                            'BodyHTML = BodyHTML & cp.Html.Hidden("AOCnt", RowPtr)
                            'Call main.AddLiveTabEntry("<NOBR>Collection&nbsp;Library</NOBR>", BodyHTML, "ccAdminTab")
                        End If
                        '
                        ' --------------------------------------------------------------------------------
                        ' Build Page from tabs
                        ' --------------------------------------------------------------------------------
                        '
                        form.addFormButton(ButtonCancel)
                        form.addFormButton(ButtonOK)
                    End If
                    '
                    ' Output the Add-on
                    '
                    form.title = "Add-on Library"
                    form.description = ""
                    If Not DbUpToDate Then
                        form.description = form.description & "<div style=""Margin-left:50px"">The Add-on Manager is disabled because this site's Database needs to be upgraded.</div>"
                    End If
                    If status <> "" Then
                        form.description = form.description & "<div style=""Margin-left:50px"">" & status & "</div>"
                    End If
                    returnResult = form.getHtml(cp)
                    ' returnResult = AdminUI.GetBody(main, Caption, ButtonList, "", False, False, Description, "", 0, Content.Text)
                    'Call main.AddPageTitle("Add-on Manager")
                End If
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
            End Try
            Return returnResult
        End Function
    End Class
End Namespace

File: source/addonManager51/Addons/CollectionLibraryClass.cs
using Contensive.BaseClasses;
using Contensive.Models.Db;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Collections.Generic;
using System.Net;
using System.Runtime.CompilerServices;

namespace Contensive.Addons.AddonManager51 {
    /// <summary>
    /// This is the Add-on Library. 
    /// </summary>
    public class CollectionLibraryClass : AddonBaseClass {
        //
        private CPBaseClass cp;
        // 
        // =====================================================================================
        /// <summary>
        /// addon api
        /// </summary>
        /// <param name="CP"></param>
        /// <returns></returns>
        public override object Execute(CPBaseClass CP) {
            try {
                cp = CP;
                return getCollectionLibrary();
            } catch (Exception ex) {
                CP.Site.ErrorReport(ex);
                throw;
            }
        }
        // 
        // ==========================================================================================================================================
        /// <summary>
        /// Addon Manager, This is a form that lets you upload an addon, Eventually, this should be substituted with a "Addon Manager Addon" - so the interface can be improved with Contensive recompile
        /// </summary>
        /// <returns></returns>
        private string getCollectionLibrary() {
            try {
                var form = cp.AdminUI.CreateLayoutBuilder();
                // 
                string SiteKey = cp.Site.GetText("sitekey", "");
                if (string.IsNullOrEmpty(SiteKey)) {
                    SiteKey = cp.Utils.CreateGuid();
                    cp.Site.SetProperty("sitekey", SiteKey);
                }
                // 
                string Button = cp.Doc.GetText(_Constants.RequestNameButton);
                string ButtonRemove = cp.Doc.GetText(_Constants.ButtonRemove);
                if ((Button ?? "") == _Constants.ButtonCancel) {
                    // 
                    // ----- redirect back to the root
                    // 
                    cp.Response.Redirect(cp.Site.GetText("adminUrl"));
                    return "";
                }
                string status = "";
                if (!cp.User.IsAdmin) {
                    return cp.Html.p("You must be an administrator to use this tool.");
                } else {
                    string installFolder = "CollectionUpload" + cp.Utils.CreateGuid().Replace("{", "").Replace("-", "").Replace("}", "");
                    string InstallPath = installFolder + @"\";
                    string ErrorMessage = "";
                    bool UpgradeOK;
                    string CollectionGUID = "";
                    if (!string.IsNullOrEmpty(ButtonRemove ?? "")) {
                        //
                        // -- uninstall
                        var collection = DbBaseModel.create<AddonCollectionModel>(cp, ButtonRemove);
                        InstallController.UninstallCollection(cp, collection.id);
                    } else if ((Button ?? "") == _Constants.ButtonOK) {
                        //
                        // -- legacy
                        // 
                        // ---------------------------------------------------------------------------------------------
                        // Download and install Collections from the Collection Library
                        // ---------------------------------------------------------------------------------------------
                        // 
                        string InstallLibCollectionList = "";
                        int Ptr;
                        if (!string.IsNullOrEmpty(cp.Doc.GetText("LibraryRow"))) {
                            Ptr = cp.Doc.GetInteger("LibraryRow");
                            CollectionGUID = cp.Doc.GetText("LibraryRowguid" + Ptr);
                            InstallLibCollectionList = InstallLibCollectionList + "," + CollectionGUID;
                        }
                        // 
                        if (!string.IsNullOrEmpty(InstallLibCollectionList)) {
                            InstallLibCollectionList = Strings.Mid(InstallLibCollectionList, 2);
                            string[] LibGuids = Strings.Split(InstallLibCollectionList, ",");
                            int cnt = Information.UBound(LibGuids) + 1;
                            var loopTo = cnt - 1;
                            for (Ptr = 0; Ptr <= loopTo; Ptr++) {
                                // 
                                // -- use v5 method
                                UpgradeOK = InstallController.installCollectionFromLibrary(cp, LibGuids[Ptr], ref ErrorMessage);
                                if (!UpgradeOK) {
                                    form.failMessage += ErrorMessage;
                                    continue;
                                }
                                if (!string.IsNullOrEmpty(ErrorMessage)) {
                                    form.warningMessage += ErrorMessage;
                                    continue;
                                }
                            }
                        }
                    } else if (!string.IsNullOrEmpty(Button ?? "")) {
                        //
                        // -- button value is the guid of the collection to install
                        // 
                        // -- use v5 method
                        UpgradeOK = InstallController.installCollectionFromLibrary(cp, Button, ref ErrorMessage);
                        if (!UpgradeOK) {
                            form.failMessage += ErrorMessage;
                        } else if (!string.IsNullOrEmpty(ErrorMessage)) {
                            form.warningMessage += ErrorMessage;
                        }
                    }


                    // 
                    ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
                    System.Xml.XmlDocument LibCollections = new System.Xml.XmlDocument();
                    LibCollections.Load("https://support.contensive.com/GetCollectionList?iv=" + cp.Version + "&key=" + cp.Utils.EncodeRequestVariable(SiteKey) + "&name=" + cp.Utils.EncodeRequestVariable(cp.Site.Name) + "&primaryDomain=" + cp.Utils.EncodeRequestVariable(cp.Site.DomainPrimary));
                    //
                    LibraryViewModel viewModel = new LibraryViewModel();
                    if ((LibCollections.DocumentElement.Name.ToLower() ?? "") != (_Constants.CollectionListRootNode.ToLower() ?? "")) {
                        string UserError = "There was an error reading the Collection Library file. The '" + _Constants.CollectionListRootNode + "' element was not found.";
                        status = status + "<BR>" + UserError;
                        cp.UserError.Add(UserError);
                    } else {
                        var RowPtr = default(int);
                        // 
                        // Go through file to validate the XML, and build status message -- since service process can not communicate to user
                        // 
                        RowPtr = 0;
                        // Content = ""
                        string cellTemplate = My.Resources.Resources.AddonManagerLibraryListCell;

                        foreach (System.Xml.XmlNode CDef_Node in LibCollections.DocumentElement.ChildNodes) {
                            string CollectionImageLink = "";
                            string CollectionCheckbox = "";
                            string CollectionName = "";
                            var CollectionModifiedDate = DateTime.MinValue;
                            string CollectionModifiedDateCaption = "";
                            string CollectionDescription = "";
                            string CollectionHelpLink = "";
                            string CollectionDemoLink = "";
                            string CollectionContensiveVersion = "";
                            switch (Strings.LCase(CDef_Node.Name) ?? "") {
                                case "collection": {
                                        // 
                                        // Read the collection
                                        // 
                                        foreach (System.Xml.XmlNode CollectionNode in CDef_Node.ChildNodes) {
                                            switch (CollectionNode.Name.ToLower() ?? "") {
                                                case "name": {
                                                        // 
                                                        // Name
                                                        // 
                                                        CollectionName = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "helplink": {
                                                        // 
                                                        // helpLink
                                                        // 
                                                        CollectionHelpLink = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "demolink": {
                                                        // 
                                                        // demoLink
                                                        // 
                                                        CollectionDemoLink = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "guid": {
                                                        // 
                                                        // Guid
                                                        // 
                                                        CollectionGUID = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "version": {
                                                        // 
                                                        // Version
                                                        // 
                                                        string CollectionVersion = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "description": {
                                                        // 
                                                        // Version
                                                        // 
                                                        CollectionDescription = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "imagelink": {
                                                        // 
                                                        // Version
                                                        // 
                                                        CollectionImageLink = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "contensiveversion": {
                                                        //
                                                        // Minimum Contensive version required
                                                        //
                                                        CollectionContensiveVersion = CollectionNode.InnerText;
                                                        break;
                                                    }
                                                case "lastchangedate": {
                                                        break;
                                                    }
                                                // 
                                                // last change - legacy field used by 4.1 to auto install, no longer updated
                                                // 
                                                case "lastmodifieddate": {
                                                        // 
                                                        // last modified
                                                        // 
                                                        CollectionModifiedDate = DateTime.MinValue;
                                                        if (Information.IsDate(CollectionNode.InnerText)) {
                                                            CollectionModifiedDate = Conversions.ToDate(CollectionNode.InnerText);
                                                        }
                                                        if (CollectionModifiedDate <= DateTime.MinValue) {
                                                            CollectionModifiedDateCaption = "unknown";
                                                        } else {
                                                            CollectionModifiedDateCaption = CollectionModifiedDate.Date.ToShortDateString();
                                                        }

                                                        break;
                                                    }
                                            }
                                        }
                                        if (string.IsNullOrEmpty(CollectionImageLink)) {
                                            CollectionImageLink = "/addonManager/libraryNoImage.jpg";
                                        }
                                        if (string.IsNullOrEmpty(CollectionModifiedDateCaption)) {
                                            CollectionModifiedDateCaption = "unknown";
                                        }
                                        if (string.IsNullOrEmpty(CollectionDescription)) {
                                            CollectionDescription = "No description is available for this add-on collection.";
                                        }
                                        bool showAddon = false;
                                        bool isInstall = false;
                                        bool isUpgrade = false;
                                        bool isRepair = false;
                                        string installedDateString = "";
                                        if (!string.IsNullOrEmpty(CollectionName)) {
                                            if (!string.IsNullOrEmpty(CollectionGUID)) {
                                                var cs = cp.CSNew();
                                                var installedDate = DateTime.MinValue;
                                                bool IsOnSite = false;
                                                if (cs.Open("Add-on Collections", "ccguid=" + cp.Db.EncodeSQLText(CollectionGUID), "", true, "ID,ModifiedDate")) {
                                                    installedDate = cs.GetDate("ModifiedDate");
                                                    installedDateString = installedDate.ToShortDateString();
                                                    IsOnSite = true;
                                                }
                                                cs.Close();
                                                CollectionCheckbox = "<input TYPE=\"CheckBox\" NAME=\"LibraryRow\" VALUE=\"" + RowPtr + "\" onClick=\"clearLibraryRows('" + RowPtr + "');\">" + cp.Html.Hidden("LibraryRowGuid" + RowPtr, CollectionGUID) + cp.Html.Hidden("LibraryRowName" + RowPtr, CollectionName);
                                                if (!IsOnSite) {
                                                    // 
                                                    // -- not installed
                                                    isInstall = true;
                                                    showAddon = true;
                                                    CollectionCheckbox += "&nbsp;Install";
                                                } else if (installedDate >= CollectionModifiedDate) {
                                                    // 
                                                    // -- up to date, reinstall
                                                    isRepair = true;
                                                    showAddon = true;
                                                    CollectionCheckbox += "&nbsp;Reinstall";
                                                } else {
                                                    // 
                                                    // -- old version, upgrade
                                                    isUpgrade = true;
                                                    showAddon = true;
                                                    CollectionCheckbox += "&nbsp;Upgrade";
                                                }
                                            }
                                        }
                                        if (!string.IsNullOrEmpty(CollectionDemoLink)) {
                                            CollectionDescription = CollectionDescription + "<div class=\"amDemoLink\"><a target=\"_blank\" href=\"" + CollectionDemoLink + "\">Demo</a></div>";
                                        }
                                        if (!string.IsNullOrEmpty(CollectionHelpLink)) {
                                            CollectionDescription = CollectionDescription + "<div class=\"amHelpLink\"><a target=\"_blank\" href=\"" + CollectionHelpLink + "\">Reference</a></div>";
                                        }
                                        bool requiresNewerContensive = false;
                                        if (!string.IsNullOrEmpty(CollectionContensiveVersion)) {
                                            try {
                                                var requiredVersion = new Version(CollectionContensiveVersion);
                                                var currentVersion = new Version(cp.Version);
                                                requiresNewerContensive = requiredVersion > currentVersion;
                                            } catch (Exception) {
                                                // -- ignore version parse errors
                                            }
                                        }
                                        if (showAddon) {
                                            viewModel.collectionList.Add(new LibraryViewModel_collectionList() {
                                                name = CollectionName,
                                                imageLink = CollectionImageLink,
                                                checkbox = CollectionCheckbox,
                                                lastUpdatedString = CollectionModifiedDateCaption,
                                                description = CollectionDescription,
                                                isInstall = isInstall,
                                                isUpgrade = isUpgrade,
                                                isRepair = isRepair,
                                                buttonValue = CollectionGUID,
                                                installedDate = installedDateString,
                                                requiresNewerContensive = requiresNewerContensive
                                            });
                                        }
                                        RowPtr += 1;
                                        break;
                                    }
                            }
                        }
                    }
                    string layout = cp.Layout.GetLayout(_Constants.guidAddonManagerLibraryListCell, _Constants.nameAddonManagerLibraryLisCell, _Constants.pathFilenameAddonManagerLibraryLisCell);
                    form.body = cp.Mustache.Render(layout, viewModel);
                    // 
                    // --------------------------------------------------------------------------------
                    // Build Page from tabs
                    // --------------------------------------------------------------------------------
                    // 
                    form.addFormButton(_Constants.ButtonCancel);
                    form.addFormButton(_Constants.ButtonOK);
                }
                // 
                // Output the Add-on
                // 
                form.title = "Add-on Library";
                form.description = "";
                if (!string.IsNullOrEmpty(status)) {
                    form.description = form.description + "<div style=\"Margin-left:50px\">" + status + "</div>";
                }
                return form.getHtml();
            } catch (Exception ex) {
                cp.Site.ErrorReport(ex);
                throw;
            }
        }
    }
    /// <summary>
    /// rendering class for the Add-on Library
    /// </summary>
    public class LibraryViewModel {
        /// <summary>
        /// list of collections
        /// </summary>
        public List<LibraryViewModel_collectionList> collectionList { get; set; } = new List<LibraryViewModel_collectionList>();
    }
    /// <summary>
    /// a collection entry in the library list
    /// </summary>
    public class LibraryViewModel_collectionList {
        /// <summary>
        /// collection name
        /// </summary>
        public string name { get; set; }
        /// <summary>
        /// image
        /// </summary>
        public string imageLink { get; set; }
        /// <summary>
        /// checkbox (should be in the form)
        /// </summary>
        public string checkbox { get; set; }
        /// <summary>
        /// string that represents the date updated
        /// </summary>
        public string lastUpdatedString { get; set; }
        /// <summary>
        /// description
        /// </summary>
        public string description { get; set; }
        /// <summary>
        /// it is not currently installed
        /// </summary>
        public bool isInstall { get; set; }
        /// <summary>
        /// a previous version is installed
        /// </summary>
        public bool isUpgrade { get; set; }
        /// <summary>
        /// it is installed and up to date. Click to reinstall/Repair
        /// </summary>
        public bool isRepair { get; set; }
        /// <summary>
        /// The button name is 'button'
        /// This is the value of the button pressed. It will be the collectionId
        /// </summary>
        public string buttonValue { get; set; } = "";
        /// <summary>
        /// if isupgrade or isrepair, this is the date it was installed
        /// </summary>
        public string installedDate { get; set; } = "";
        /// <summary>
        /// true if the collection requires a newer version of Contensive than is currently installed
        /// </summary>
        public bool requiresNewerContensive { get; set; }

    }
}